### PR TITLE
Fix dependency resolving

### LINKED ISSUE #8365: synapse fails to start if wrong version of isort is installed
This seems to be a recent regression.

Why does having the wrong version of isort installed prevent synapse from starting? Synapse doesn't use isort at runtime, so surely this makes no sense?

```
(env3)rav@fred:~/work/synapse (develop *$%=)$ python -m synapse.app.homeserver
ERROR:root:Needed optional isort==5.0.3, got isort==5.0.5
Missing Requirements: 'isort==5.0.3'
To install run:
    pip install --upgrade --force 'isort==5.0.3'
```

### PR DESCRIPTION
**Pull request by ShadowJonathan**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Fixes #8365

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

File: synapse/python_dependencies.py [hunk 1]
```diff
@@ -221,3 +230,5 @@ def _check_requirement(dependency_string):
     import sys
 
     sys.stdout.writelines(req + "\n" for req in list_requirements())
+
+    check_requirements(dev=True)
```
Comment (clokep) [inline synapse/python_dependencies.py:234]: Does something run this module at some point? When is this used?

File: changelog.d/8367.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Fixes the way synapse distinguishes between runtime and developer dependencies.
```
Comment (clokep) [inline changelog.d/8367.misc:1]: We should use the same newsfragment as the PR that regressed this so that they'll show up together in the changelog.

File: synapse/python_dependencies.py [hunk 3]
```diff
@@ -139,12 +142,12 @@ def dependencies(self):
...
 
 
-def check_requirements(for_feature=None):
+def check_requirements(for_feature=None, dev=False):
     deps_needed = []
     errors = []
 
     if for_feature:
-        reqs = CONDITIONAL_REQUIREMENTS[for_feature]
+        reqs = CONDITIONAL_REQUIREMENTS.get(for_feature, [])
```
Comment (clokep) [inline synapse/python_dependencies.py:150]: Why this change? It seems unrelated?

**Review by clokep — CHANGES_REQUESTED** [review]

**Review by richvdh — COMMENTED** [review]
generally this feels overengineered to me. I don't really see why we need to complicate `python_dependencies` with this stuff rather than just putting the dev dependencies in setup.py.

File: synapse/python_dependencies.py (hunk 3 above)
Comment (ShadowJonathan) [reply synapse/python_dependencies.py:150]: if `for_feature` is not a key in `CONDITIONAL_REQUIREMENTS`, it'll raise a `KeyError`, i think the right logic then is to return a default empty list

This is so that when `check_requirements` is called with a dependency list that does not exist anymore, any time in the future, it should not return a cryptic error, should i revert this?

File: synapse/python_dependencies.py (hunk 1 above)
Comment (ShadowJonathan) [reply synapse/python_dependencies.py:234]: No, currently nothing runs the module at this point, but this way, `python -m synapse.python_dependencies` can be suggested by someone from the synapse team to be ran by developers to check if their dependencies are alright, or if one of them is the wrong version.

**Comment by ShadowJonathan** [comment]
@richvdh it's because i feel this fixes some other problems besides #8365

**Comment by richvdh** [comment]
> @richvdh it's because i feel this fixes some other problems besides #8365

well, perhaps you could clarify what they are. I feel like we've discussed this, asked you to take a different approach, and now you've ignored our requests and presented us with your own implementation without explaining how it works and why it is superior.

File: synapse/python_dependencies.py [hunk 4]
```diff
@@ -116,6 +113,12 @@
     if name not in ["systemd", "lint"]:
         ALL_OPTIONAL_REQUIREMENTS = set(optional_deps) | ALL_OPTIONAL_REQUIREMENTS
 
+# We exclude lint and test as those are developer requirements
+ALL_RUNTIME_REQUIREMENT_COMPONENTS = set(CONDITIONAL_REQUIREMENTS.keys()) - {
+    "lint",
+    "test",
+}
+
```
Comment (ShadowJonathan) [inline synapse/python_dependencies.py:121]: @richvdh This change filters out all runtime dependencies from all conditional dependencies, by only excluding `lint` and `test`, and then flattening the resulting dependencies into a single `set`.

I could change this to be only a set with `{"lint", "test"}`, as this piece is only used in one part later on.

File: synapse/python_dependencies.py [hunk 5]
```diff
@@ -171,10 +174,16 @@ def check_requirements(for_feature=None):
...
+        # Check if the optional runtime dependencies are up to date. We allow them to not be
         # installed.
-        OPTS = sum(CONDITIONAL_REQUIREMENTS.values(), [])  # type: List[str]
+        OPTS = set().union(
+            *(
+                set(reqs)
+                for key, reqs in CONDITIONAL_REQUIREMENTS.items()
+                if key in ALL_RUNTIME_REQUIREMENT_COMPONENTS or dev
+            )
+        )  # type: Set[str]
```
Comment (ShadowJonathan) [inline synapse/python_dependencies.py:186]: @richvdh this is two compounded changes;

- Change `OPTS` to be a flattened set, to deduplicate any elements, and have potential faster iteration.
- if `dev == False`, only check default (runtime) dependencies, if `dev == True`, also check development dependencies, this is useful as an addition in `__main__` down below, which does `dev=True`, making an invocation of the file (`python -m synapse.python_dependencies`) automatically also check those development dependencies.

**Comment by ShadowJonathan** [comment]
> > @richvdh it's because i feel this fixes some other problems besides #8365
> 
> well, perhaps you could clarify what they are. I feel like we've discussed this, asked you to take a different approach, and now you've ignored our requests and presented us with your own implementation without explaining how it works and why it is superior.

I left some comments doing exactly that last part, sorry for my abrasive changes, I thought this was a rational change overall.

**Comment by ShadowJonathan** [comment]
Some extra context, `setup.py` does

https://github.com/matrix-org/synapse/blob/b29a9bdaa9954adb0ae6289fd278572100dabef1/setup.py#L94-L95

https://github.com/matrix-org/synapse/blob/b29a9bdaa9954adb0ae6289fd278572100dabef1/setup.py#L103-L104

Adding the following to it would split the concern of "dependency notation" between `synapse/python_dependencies.py` and `setup.py`, i think adding this to `setup.py` when nearly everything is already being defined in `synapse.python_dependencies` would lead to some (more) loss of oversight in the matter.

```py
# Make `pip install matrix-synapse[all]` install all the optional dependencies.
CONDITIONAL_REQUIREMENTS["all"] = list(ALL_OPTIONAL_REQUIREMENTS)

CONDITIONAL_REQUIREMENTS["lint"] = ["isort==5.0.3", "black==19.10b0", "flake8-comprehensions", "flake8"]
```

The synapse codebase wants to define dependencies in `synapse.python_dependencies`, that's fine, but i think then that invisible agreement should not be breached by adding extra (dependent (because of `tox` installing dependencies because of it)) behaviour in a different spot, effectively having to have someone look up in multiple places if some kind of bug was present here.

**Review by richvdh — CHANGES_REQUESTED** [review]
ok so to summarise, I think you are saying you want to put it in `python_dependencies` because that is where existing dependencies are listed and you don't want to split them up.

my counterpoint is: this complicates `python_dependencies`, making it harder to understand. Additionally I would say that `python_dependencies` is explicitly for *runtime* dependencies, and having other dependencies elsewhere is reasonable.

I don't particularly want to debate this any further. If you'd like to update this PR to change `setup.py` instead, that would be great. Otherwise I can revert the PR that caused the regression. Thank you.

**Comment by clokep** [comment]
I think that #8377 is what @richvdh is asking for.
